# Individual Database Profiling Report: `tmp_benchmark_wide_text_nulls`

--- 
## 1. Introduction

This document serves as the standardized Data Profiling Report for the `tmp_benchmark_wide_text_nulls` database. This database is a critical component of Phase 1 of the Digital TMP project, an initiative dedicated to the systematic, quantitative evaluation of legacy and modern database architectures to inform a strategic redesign. The purpose of this report is to provide a deep-dive analysis of this specific benchmark schema by visualizing and interpreting a suite of pre-computed metrics. By presenting granular, empirical evidence in a structured format, this report establishes a baseline for understanding the structural simplicity, data integrity, and performance characteristics of a fully denormalized and modernized data model.

This report is a visualization and interpretation layer for a standardized set of metrics generated by an automated data profiling pipeline, as defined in the Phase 1 project plan. It does not represent a live analysis but rather a reproducible summary of the database's state. The analysis will proceed systematically, examining the database's schema, table-level metrics, column-level data profiles, and performance on canonical analytical queries. The findings presented herein are crucial; this benchmark represents the project's ideal target architecture, combining the performance benefits of denormalization with modern, user-friendly data representation. Its profile will be central to the final Phase 1 White Paper, providing the definitive, evidence-based argument for the recommended Phase 2 design.

--- 
## 2. Background

### 2.1. Context and Motivation

The primary goal of Phase 1 of the Digital TMP project is to conduct a systematic and quantitative evaluation of four legacy databases (`TMP_DF8`, `TMP_DF9`, `TMP_DF10`, `TMP_REAN_DF2`) and two modern benchmark databases. As outlined in the project's architectural and planning documents, this evaluation is a foundational step designed to generate the empirical evidence required to inform a strategic architectural redesign in Phase 2. The existing legacy databases, developed over several decades, exhibit a range of structural complexities and performance characteristics that must be rigorously measured and compared before a new, unified schema can be designed.

This report, focused specifically on `tmp_benchmark_wide_text_nulls`, is one of six standardized analyses. Its role is unique: it does not analyze a historical artifact but a purpose-built test case representing the project's proposed ideal. It was created to serve as the "gold standard" against which all other schemas are measured. By applying the same rigorous profiling to this modernized schema as to the complex legacy ones, the project can isolate and quantify the precise impact of architectural choices—normalization vs. denormalization—as well as data representation choices—numeric codes vs. human-readable text and sentinel values vs. standard NULLs.

### 2.2. Data: The `tmp_benchmark_wide_text_nulls` Database

The `tmp_benchmark_wide_text_nulls` database is a modern, purpose-built database created specifically for this analytical phase. Its provenance is defined by the Extract, Transform, Load (ETL) logic in the `phases/01_LegacyDB/sql/flatten_df9_text_nulls.sql` script. This script flattens the highly normalized `TMP_DF9` database into a single wide table and, crucially, implements two modernizing transformations that set it apart from its twin, the `numeric` benchmark.

The content of this benchmark database comprises the complete set of attributes for 5,050 archaeological sites, identical to that of its source, `TMP_DF9`. Structurally, it is a radical simplification, consisting of a **single table** (`wide_format_data`) with **337 columns**, thereby completely eliminating the need for relational joins to assemble a site record. Its two defining features, implemented by the ETL script, are:
1.  **Conversion to Human-Readable Text:** Unlike the legacy databases and the `numeric` benchmark, this schema replaces opaque integer codes with their corresponding human-readable text descriptions. This is achieved by joining `TMP_DF9`'s core data tables with its 45 "Codes" lookup tables during the ETL process. The resulting table contains clear, descriptive values (e.g., `"1. Low"` instead of just `1`), significantly enhancing usability and interpretability for the end user.
2.  **Conversion to Standard SQL NULLs:** This benchmark is the only database in the Phase 1 analysis that correctly handles missing data. The ETL script explicitly converts `TMP_DF9`'s non-standard sentinel values (e.g., `-1`, `-999`) into standard SQL `NULL`s. This is a critical data integrity improvement that prevents the corruption of statistical calculations and aligns the database with modern best practices.

In sum, this database was designed to represent the ideal target architecture for Phase 2: a structurally simple, high-performance, and analytically intuitive model that is both human-readable and uses standard conventions for data integrity.

### 2.3. Methods: Database Profiling Metrics

The analysis presented in this report is based on a standardized suite of pre-computed metrics generated by the `02_run_profiling_pipeline.py` script, as outlined in the Phase 1 project plan. This methodological approach ensures that the evaluation of `tmp_benchmark_wide_text_nulls` is both reproducible and directly comparable to the analyses of the other five databases under review. The metrics were calculated from a live PostgreSQL instance of the database and saved to disk as discrete JSON and CSV files. This notebook serves as the visualization and interpretation layer for this static, pre-computed data, rather than performing a live analysis. This ensures that the findings reported here are a stable, verifiable snapshot of the database's characteristics at the time the pipeline was executed.

The profiling pipeline gathers data across several distinct categories to provide a holistic assessment of the database. The report will present findings from each of these categories in sequence. **Schema-level metrics** provide a high-level overview, including aggregate object counts (e.g., table count) and total database size. **Table-level metrics** offer a more granular view of the single table, assessing its size, row count, and health. **Column-level analysis** provides the deepest insights, examining both the structure (data types, nullability) and content (NULL value percentages, cardinality) of every column. Finally, this report presents **performance benchmarks** that measure query latency on a set of canonical analytical workloads.

### 2.4. Hypotheses

Based on its purpose-built, denormalized, and modernized design, this analysis is guided by a two-part hypothesis for the `tmp_benchmark_wide_text_nulls` database:
1.  **Performance:** It is hypothesized that its single-table, "wide-format" architecture will demonstrate exceptional performance on the analytical queries defined in `canonical_queries_benchmark.sql`. We predict its query latencies will be extremely low and comparable to its `numeric` twin, providing powerful, direct evidence for the performance benefits of a denormalized structure.
2.  **Usability & Data Integrity:** It is further hypothesized that its data model, which employs human-readable `text` for categorical data and standard SQL `NULL`s for missing values, represents a superior design for usability and data integrity compared to *all* other databases in this analysis. The column-level metrics are expected to validate that this model eliminates the analytical risks and cognitive overhead associated with numeric codes and non-standard sentinel values.

---
## 3. Setup and Configuration

In [ ]:
import json
from pathlib import Path

import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from IPython.display import SVG, Markdown, display
from plotly.subplots import make_subplots

# --- CONFIGURATION ---------------------------------------------------
# SET THIS VARIABLE to the name of the database you want to analyze.
# e.g., 'tmp_df8', 'tmp_df9', 'tmp_benchmark_wide_numeric', etc.
DATABASE_NAME = "tmp_benchmark_wide_text_nulls"  # <--- CHANGE THIS
# ---------------------------------------------------------------------

# --- Path Definitions ---
# Use relative paths from the notebook's location in reports/individual_db_analysis/
METRICS_DIR = Path("../../outputs/metrics")
ERDS_DIR = Path("../../outputs/erds")

# --- Styling and Display Options ---
pd.set_option("display.max_columns", 100)
pd.set_option("display.width", 120)


def display_header(title):
    display(Markdown(f"### {title}"))


def load_metric_file(metric_name, file_type="csv"):
    """Helper function to safely load a metric file."""
    file_path = METRICS_DIR / f"{DATABASE_NAME}_{metric_name}.{file_type}"
    if not file_path.exists():
        print(f"⚠️ WARNING: Metric file not found: {file_path.name}")
        return None
    if file_type == "csv":
        return pd.read_csv(file_path)
    elif file_type == "json":
        with open(file_path, "r") as f:
            return json.load(f)


print(f"✅ Setup complete. Analyzing database: '{DATABASE_NAME}'")
print(f"Metrics Directory: {METRICS_DIR}")
print(f"ERD Directory: {ERDS_DIR}")

✅ Setup complete. Analyzing database: 'tmp_benchmark_wide_text_nulls'
Metrics Directory: ..\..\outputs\metrics
ERD Directory: ..\..\outputs\erds


---
## 4. Data Loading

In [ ]:
# Load all metric files into variables
basic_metrics = load_metric_file("basic_metrics", "json")
schema_counts = load_metric_file("schema_counts", "json")
# Note: interop_metrics file doesn't exist for benchmark databases
interop_metrics = None

# Load table metrics and convert to DataFrame
table_metrics_data = load_metric_file("table_metrics", "json")
table_metrics_df = pd.DataFrame(table_metrics_data) if table_metrics_data else None

# Load column structure and convert to DataFrame
column_structure_data = load_metric_file("column_structure", "json")
column_structure_df = (
    pd.DataFrame(column_structure_data) if column_structure_data else None
)

# Load column profiles and convert to DataFrame
column_profiles_data = load_metric_file("column_profiles", "json")
column_profiles_df = (
    pd.DataFrame(column_profiles_data) if column_profiles_data else None
)

# Performance benchmarks remain as CSV
performance_df = load_metric_file("performance_benchmarks")

print("✅ Data loading complete.")

✅ Data loading complete.


---
## 5. High-Level Overview & Schema Visualization

### 5.1. Data: Database and Schema-Level Metrics

This section presents a high-level, aggregate analysis of the `tmp_benchmark_wide_text_nulls` database, providing a "30,000-foot view" of its overall size and structural complexity. The metrics presented below are schema-wide statistics, sourced from the `tmp_benchmark_wide_text_nulls_basic_metrics.json` and `tmp_benchmark_wide_text_nulls_schema_counts.json` files generated by the profiling pipeline. These summary statistics offer a quantitative starting point for assessing the database's architecture. As this benchmark database is designed to be non-relational, the custom interoperability metrics (JDI, LIF, NF) were not calculated and are not applicable.

The summary table will present key metrics that quantify the database's scale. `table_count` is a direct measure of structural fragmentation, and `database_size_mb` quantifies the total disk space consumed by the database.

### 5.2. Theory & Methods: Schema Complexity and Interoperability Metrics

The schema-level metrics presented in this section are derived from two sources: direct queries against the PostgreSQL `information_schema` catalog and a suite of custom heuristic scores designed to quantify architectural complexity. The combination of these methods provides a multi-faceted, quantitative assessment of the database schema.

**Schema Object Counts:** Fundamental metrics such as `table_count`, `view_count`, and `function_count` are derived from straightforward `COUNT` queries on the relevant tables within the `information_schema`. For example, the table count is obtained by querying `information_schema.tables` where the `table_schema` matches the target schema of the analysis. These direct counts provide a baseline measure of the number of discrete objects that comprise the database structure.

**Interoperability Metrics:** To move beyond simple counts, three custom heuristic metrics were developed to provide a more nuanced assessment of relational complexity. These are defined as follows:
*   **Join Dependency Index (JDI):** The JDI is a measure of relational complexity based on the density of defined foreign key relationships. It is calculated as `JDI = foreign_key_count / max_possible_foreign_keys`, where `max_possible_foreign_keys` is derived from the number of tables (`n`) as `n * (n - 1) / 2`. A JDI score closer to 1.0 indicates a schema with a dense web of explicit relationships, suggesting higher normalization, while a score closer to 0 indicates a schema with few formal relationships, typical of denormalized or fragmented designs.
*   **Logical Interoperability Factor (LIF):** The LIF is a heuristic designed to estimate the potential for *ad hoc* joins where formal foreign keys may not exist. It operates on the assumption that columns with similar names and data types across different tables are likely to represent the same logical entity and are thus joinable. The metric is calculated by counting the number of distinct column name/data type pairs that appear in more than one table, providing a rough measure of logical cohesion.
*   **Normalization Factor (NF):** The NF is a composite score that combines the `table_count` and the `JDI` into a single, normalized value between 0 and 1. This score provides a holistic heuristic for the degree of schema normalization or fragmentation. Higher values suggest a more normalized schema (many tables with dense relationships), while lower values indicate a simpler, denormalized, or fragmented structure. The primary value of these heuristic scores lies not in their absolute values, but in their utility for *relative comparison* across the different database schemas analyzed in Phase 1.

In [ ]:
display_header(f"Key Metrics for: {DATABASE_NAME}")

summary_data = {}
if basic_metrics:
    summary_data.update(basic_metrics)
if schema_counts:
    summary_data.update(schema_counts)
if interop_metrics:
    summary_data.update(interop_metrics)
if table_metrics_df is not None:
    summary_data["total_estimated_rows"] = int(table_metrics_df["row_estimate"].sum())

if summary_data:
    summary_series = pd.Series(summary_data).rename("Value").to_frame()
    display(summary_series)
else:
    print("No summary metrics available.")

### Key Metrics for: tmp_benchmark_wide_text_nulls

### 5.3. Results: Key Metrics for `tmp_benchmark_wide_text_nulls`

The high-level metrics for `tmp_benchmark_wide_text_nulls` confirm its intended design as a structurally simple but content-rich schema. The database consists of a **single table** and has no views, functions, or sequences. It occupies **62 MB** of disk space, making it the second-largest database in the analysis, a direct result of replacing efficient integer codes with more descriptive but larger text fields. As there are no foreign key relationships and only one table, the interoperability and normalization metrics are effectively zero, representing the lowest possible level of relational complexity.

### 5.4. Discussion: Initial Assessment of Schema Complexity

The quantitative schema-level metrics provide decisive support for the hypothesis that the `tmp_benchmark_wide_text_nulls` database embodies a simple and structurally optimized design. With a `table_count` of one and complexity scores of zero, the schema represents the logical endpoint of denormalization. This extreme simplicity is a deliberate architectural choice, designed to entirely eliminate the need for performance-intensive join operations at query time. The larger database size compared to its `numeric` twin (62 MB vs. 21 MB) is the expected and accepted trade-off for enhanced usability and interpretability through the use of human-readable text. This initial assessment establishes the benchmark database as the ideal case for both performance and data integrity against which the legacy schemas can be quantitatively measured.

### 5.5. Data & Methods: Entity-Relationship Diagram (ERD)

An Entity-Relationship Diagram (ERD) is a visual representation of a database schema that illustrates the entities (tables), their attributes (columns), and the relationships between them. It serves as a critical tool for understanding the logical structure of a database, revealing its degree of normalization, the nature of its data dependencies, and its overall architectural complexity.

The ERD presented in this report was not manually drawn but was generated through an automated process to ensure it provides a completely accurate and objective reflection of the live `tmp_benchmark_wide_text_nulls` database schema. The diagram was created by the `03_generate_erds.py` script, which uses the `sqlalchemy-schemadisplay` library to programmatically inspect the live database's metadata. This process automatically discovers all tables, columns, and foreign key constraints and uses the `graphviz` software toolkit to render a graphical representation of these objects and their relationships. This automated methodology guarantees that the ERD is a direct, empirical visualization of the database's actual structure, free from any idealization or interpretation.

In [ ]:
display_header(f"Full ERD for: {DATABASE_NAME}")

try:
    # Find the most recent ERD file for the database
    erd_files = sorted(ERDS_DIR.glob(f"{DATABASE_NAME}_full_ERD_*.svg"), reverse=True)
    if erd_files:
        display(SVG(erd_files[0]))
    else:
        print(f"❌ ERROR: Full ERD SVG file not found for '{DATABASE_NAME}'.")
except Exception as e:
    print(f"An error occurred while displaying the ERD: {e}")

### Full ERD for: tmp_benchmark_wide_text_nulls

### 5.6. Results & Discussion: Visualizing Relational Complexity

The Entity-Relationship Diagram for `tmp_benchmark_wide_text_nulls` provides the most direct and intuitive visual evidence of its architectural simplicity. The diagram consists of a single, isolated box representing the `wide_format_data` table. There are no other tables and, consequently, no relationship lines.

This minimalist visualization perfectly corroborates the quantitative metrics, translating a `table_count` of one and a JDI of zero into an unambiguous graphical representation. It stands in stark and intentional contrast to the complex "spaghetti diagrams" of the legacy databases like `TMP_DF9`. The ERD serves as a powerful visual argument for the core principle of the Phase 1 redesign: by denormalizing the schema into a single wide table, the relational complexity that burdens query performance and user comprehension is completely eliminated. This diagram is the visual embodiment of an architecture optimized for analytical speed and usability.

---
## 6. Table-Level Analysis

### 6.1. Data: Table-Level Metrics

This section presents a focused analysis of the single table that constitutes the entirety of the `tmp_benchmark_wide_text_nulls` database. The data is sourced from the `tmp_benchmark_wide_text_nulls_table_metrics.json` file. By examining the metrics for the `wide_format_data` table, we can assess its scale, health, and storage characteristics.

The upcoming summary table will present key metrics for this single table. The `row_estimate` provides the number of rows, corresponding to the total number of archaeological sites. `column_count` indicates the table's width. `total_size` reports the total disk space consumed by the table. `bloat_percent` is a health indicator, quantifying the percentage of unused space within the table's file.

### 6.2. Theory & Methods: Assessing Table Health and Size

The table-level metrics presented in this section are calculated using standard PostgreSQL functions and statistical queries that provide efficient and reliable information about table size and health. The methods used are designed to avoid performance-intensive operations while still yielding accurate assessments.

**Row Estimates:** The `row_estimate` for each table is not derived from an expensive `COUNT(*)` operation, which would require a full table scan. Instead, it is a highly efficient estimate sourced directly from the `pg_class.reltuples` column in PostgreSQL's internal statistics catalog. This value is updated by the `ANALYZE` command (and autovacuum daemon) and typically provides a very close approximation of the actual row count for static or infrequently updated tables, making it a standard and performant method for assessing table size.

**Table and Index Size:** The various size metrics (`table_size`, `index_size`, `total_size`) are calculated using built-in PostgreSQL functions such as `pg_relation_size()` and `pg_total_relation_size()`. These functions measure the actual disk space allocated to the table's data file (the "heap") and its associated indexes. The values are presented in a human-readable format (e.g., kB, MB) for easier interpretation.

**Table Bloat:** Table bloat refers to unused space that accumulates within a PostgreSQL table's data file due to the database's Multi-Version Concurrency Control (MVCC) implementation. When rows are updated (`UPDATE`) or deleted (`DELETE`), the old row versions are not immediately removed from the file; they are marked as "dead" and remain until a `VACUUM` process reclaims the space. `bloat_percent` is a key indicator of database health and is calculated using a standard community-provided SQL query that statistically estimates the amount of this dead space. High bloat percentages can negatively impact performance by increasing the number of disk pages that must be scanned to satisfy a query. It often suggests a need for database maintenance, such as running a `VACUUM FULL` operation or tuning autovacuum settings.

In [ ]:
display_header("Table Metrics Summary")

if table_metrics_df is not None and not table_metrics_df.empty:
    display(
        table_metrics_df.sort_values(
            by="row_estimate", ascending=False
        ).style.background_gradient(
            cmap="viridis", subset=["row_estimate", "bloat_percent"]
        )
    )
else:
    print("No table metrics data available.")

### Table Metrics Summary

### 6.3. Results: Table Metrics Summary for `tmp_benchmark_wide_text_nulls`

The metrics for the single `wide_format_data` table confirm its status as a large, content-rich, denormalized entity. The table has a row estimate of **5,050 rows**, consistent with the number of sites in its source database, `TMP_DF9`. It is exceptionally wide, with **337 columns** that consolidate attributes from across `TMP_DF9`'s original schema.

The total on-disk size of this single table is **55 MB**. This is substantially larger than its `numeric` counterpart (21 MB), a direct and expected consequence of storing human-readable text descriptions instead of compact integer codes. The table health is poor, with an exceptionally high bloat percentage of **142.98%**.

In [ ]:
display_header("Largest Tables by Total Size and Bloat")

if table_metrics_df is not None and not table_metrics_df.empty:
    # Convert pretty size string to bytes for sorting
    def size_to_bytes(s):
        if not isinstance(s, str):
            return 0
        num, unit = s.split()
        num = float(num)
        if "KB" in unit:
            return num * 1024
        if "MB" in unit:
            return num * 1024**2
        if "GB" in unit:
            return num * 1024**3
        return num

    df_copy = table_metrics_df.copy()
    df_copy["total_bytes"] = df_copy["total_size"].apply(size_to_bytes)
    df_copy["bloat_bytes_val"] = df_copy["bloat_bytes"]

    top_10_size = df_copy.nlargest(10, "total_bytes")
    top_10_bloat = df_copy.nlargest(10, "bloat_bytes_val")

    # Display tables
    display(Markdown("**Top 10 Tables by Total Size**"))
    display(
        top_10_size[["table_name", "total_size", "row_estimate"]].reset_index(drop=True)
    )

    display(Markdown("**Top 10 Tables by Bloat Size**"))
    display(
        top_10_bloat[
            ["table_name", "bloat_size", "bloat_percent", "row_estimate"]
        ].reset_index(drop=True)
    )

    # Create subplots
    fig = make_subplots(
        rows=1,
        cols=2,
        subplot_titles=("Top 10 Tables by Total Size", "Top 10 Tables by Bloat Size"),
    )

    fig.add_trace(
        go.Bar(
            y=top_10_size["table_name"],
            x=top_10_size["total_bytes"],
            orientation="h",
            name="Total Size",
        ),
        row=1,
        col=1,
    )
    fig.add_trace(
        go.Bar(
            y=top_10_bloat["table_name"],
            x=top_10_bloat["bloat_bytes_val"],
            orientation="h",
            name="Bloat Size",
        ),
        row=1,
        col=2,
    )

    fig.update_layout(
        title_text=f"Table Size Analysis for {DATABASE_NAME}",
        height=500,
        showlegend=False,
    )
    fig.update_yaxes(autorange="reversed")
    fig.update_xaxes(title_text="Size (Bytes)", row=1, col=1)
    fig.update_xaxes(title_text="Bloat (Bytes)", row=1, col=2)
    fig.show()
else:
    print("No table metrics data available for plotting.")

### Largest Tables by Total Size and Bloat

**Top 10 Tables by Total Size**

,table_name,total_size,row_estimate
0,wide_format_data,55 MB,5050


**Top 10 Tables by Bloat Size**

,table_name,bloat_size,bloat_percent,row_estimate
0,wide_format_data,27.03 MB,142.98,5050


### 6.4. Results: Largest Tables by Size and Bloat

As there is only one table, the `wide_format_data` table is trivially the largest by both total size (55 MB) and bloat size (27.03 MB). The extremely high bloat percentage of 142.98% is the key finding. This indicates that the amount of wasted, reclaimable space within the table's data file is larger than the space occupied by the actual data itself.

### 6.5. Discussion: Identifying Key Tables and Health Concerns

The table-level analysis provides a clear profile of the `text_nulls` benchmark. It successfully consolidates all project data into a single, wide table with 5,050 rows, confirming the ETL process correctly preserved all site records. The substantial 55 MB size is the direct and anticipated trade-off for prioritizing usability; storing descriptive text is inherently less space-efficient than storing numeric codes.

The most significant finding is the exceptionally high table bloat of 142.98%. This is not an indicator of data churn or corruption but rather a likely artifact of the specific ETL process used to create the table. A complex `CREATE TABLE AS SELECT ...` operation involving numerous joins and transformations, like the `flatten_df9_text_nulls.sql` script, can lead PostgreSQL to allocate significantly more page space than is strictly necessary for the final data. For a static, read-only analytical table, this is not a major operational concern. A single `VACUUM FULL` command would reclaim this space and significantly reduce the table's on-disk footprint. The simplicity and content richness of the single-table structure remain its defining and most important features.

---
## 7. Column-Level Analysis

### 7.1. Data Type Frequencies

#### 7.1.1. Data & Methods

This analysis examines the fundamental composition of the `tmp_benchmark_wide_text_nulls` schema by profiling the data types used across all of its columns. The data for this section is sourced from the `tmp_benchmark_wide_text_nulls_column_structure.json` file, which contains metadata for every column in the database, including its assigned PostgreSQL data type. The method involves aggregating this data to count the frequency of each distinct data type. The resulting distribution provides critical insight into the database's design philosophy, particularly regarding its approach to storing categorical, numeric, and textual information.

In [ ]:
display_header("Data Type Distribution")

if column_structure_df is not None:
    type_counts = column_structure_df["data_type"].value_counts().reset_index()
    type_counts.columns = ["data_type", "count"]

    # Calculate percentages
    type_counts["percentage"] = (
        type_counts["count"] / type_counts["count"].sum() * 100
    ).round(2)

    # Display comprehensive table
    display(Markdown("**Complete Data Type Distribution**"))
    display(type_counts.style.format({"percentage": "{:.2f}%"}))

    # Display summary statistics
    display(Markdown("**Data Type Summary**"))
    summary_stats = pd.DataFrame(
        {
            "Total Columns": [type_counts["count"].sum()],
            "Unique Data Types": [len(type_counts)],
            "Most Common Type": [
                f"{type_counts.iloc[0]['data_type']} ({type_counts.iloc[0]['count']} columns)"
            ],
            "Least Common Type": [
                f"{type_counts.iloc[-1]['data_type']} ({type_counts.iloc[-1]['count']} columns)"
            ],
        }
    )
    display(summary_stats)

    fig = px.bar(
        type_counts,
        x="data_type",
        y="count",
        title=f"Column Data Type Frequencies in {DATABASE_NAME}",
        labels={"count": "Number of Columns", "data_type": "Data Type"},
    )
    fig.show()
else:
    print("No column structure data available.")

### Data Type Distribution

**Complete Data Type Distribution**

**Data Type Summary**

,Total Columns,Unique Data Types,Most Common Type,Least Common Type
0,337,4,double precision (161 columns),bigint (1 columns)


#### 7.1.2. Results & Discussion

The analysis of column data types in the `tmp_benchmark_wide_text_nulls` database reveals a schema that prioritizes human-readable text and standard numeric types. Of the 337 columns in the `wide_format_data` table, **161 (47.8%)** are `double precision`, **122 (36.2%)** are `text`, **53 (15.7%)** are `boolean`, and only **1 (0.3%)** is a `bigint`.

This distribution is a direct result of the modernizing transformations applied by the `flatten_df9_text_nulls.sql` ETL script. The most significant finding is the prevalence of the `text` data type. This confirms that the script successfully joined `TMP_DF9`'s core data with its 45 "Codes" lookup tables, converting the original, opaque `smallint` codes into meaningful, descriptive strings. The `double precision` columns represent the raw artifact counts, while the `boolean` columns are the result of pivoting personnel data. This data type profile demonstrates a clear design choice for usability and interpretability, creating a dataset that is immediately understandable without requiring external documentation or complex joins. It stands in stark contrast to the legacy schemas and the `numeric` benchmark, which rely heavily on integer codes.

### 7.2. Data Completeness: NULL Value Analysis

#### 7.2.1. Data & Methods

This analysis assesses data completeness and quality by measuring the prevalence of `NULL` values across every column in the `tmp_benchmark_wide_text_nulls` database. Sourced from the `tmp_benchmark_wide_text_nulls_column_profiles.json` file, the core metric is `null_percent`, which calculates the percentage of rows containing a `NULL` for each column. In standard SQL, `NULL` is the correct and conventional representation for missing or unknown data. This analysis is a critical measure of data integrity, as the ETL process for this specific benchmark was designed to convert legacy sentinel values into proper `NULL`s.

In [ ]:
display_header("Top 20 Columns by Percentage of NULL Values")

if column_profiles_df is not None and not column_profiles_df.empty:
    # Ensure we only show columns with NULLs
    null_df = column_profiles_df[column_profiles_df["null_percent"] > 0].copy()

    if not null_df.empty:
        # Create a full column identifier for clarity
        null_df["full_column_name"] = (
            null_df["tablename"].astype(str) + "." + null_df["column_name"].astype(str)
        )

        top_20_nulls = null_df.nlargest(20, "null_percent")

        # Display table
        display(Markdown("**Top 20 Columns with Highest NULL Percentages**"))
        table_display = top_20_nulls[
            [
                "full_column_name",
                "null_percent",
                "null_count_estimate",
                "row_count_exact",
            ]
        ].copy()
        table_display.columns = ["Column", "NULL %", "NULL Count", "Total Rows"]
        display(table_display.reset_index(drop=True))

        # Display summary statistics
        display(Markdown("**NULL Value Summary**"))
        null_summary = pd.DataFrame(
            {
                "Total Columns Analyzed": [len(column_profiles_df)],
                "Columns with NULLs": [len(null_df)],
                "Columns with 100% NULLs": [
                    len(null_df[null_df["null_percent"] == 100])
                ],
                "Average NULL %": [f"{null_df['null_percent'].mean():.2f}%"],
                "Median NULL %": [f"{null_df['null_percent'].median():.2f}%"],
            }
        )
        display(null_summary)

        fig = px.bar(
            top_20_nulls,
            y="full_column_name",
            x="null_percent",
            orientation="h",
            title=f"Top 20 Columns by NULL Percentage in {DATABASE_NAME}",
            labels={
                "null_percent": "Percentage of Rows that are NULL (%)",
                "full_column_name": "Column",
            },
        )
        fig.update_layout(height=600)
        fig.update_yaxes(autorange="reversed")
        fig.show()
    else:
        print("✅ Excellent! No columns with NULL values were found.")

        # Still show summary even when no NULLs
        display(Markdown("**NULL Value Summary**"))
        null_summary = pd.DataFrame(
            {
                "Total Columns Analyzed": [len(column_profiles_df)],
                "Columns with NULLs": [0],
                "Data Completeness": ["100% - Perfect!"],
            }
        )
        display(null_summary)
else:
    print("No column profile data available.")

### Top 20 Columns by Percentage of NULL Values

**Top 20 Columns with Highest NULL Percentages**

**NULL Value Summary**

,Total Columns Analyzed,Columns with NULLs,Columns with 100% NULLs,Average NULL %,Median NULL %
0,337,265,0,14.88%,4.59%


#### 7.2.2. Results & Discussion

The NULL value analysis of the `tmp_benchmark_wide_text_nulls` database reveals a dataset with widespread but expected patterns of missing data. The analysis found `NULL` values in **265 of the 337 columns**, with a median `NULL` percentage of **4.59%**. This result is highly significant because it demonstrates the successful conversion of non-standard sentinel values (like `-1` from `TMP_DF9`) into proper SQL `NULL`s, a key data integrity goal of this benchmark.

The columns with the highest NULL percentages are primarily rare artifact types or specialized architectural interpretations, with some exceeding 90% NULLs (e.g., `wallCoreStone`, `wallFacing`, `floorMaterial`). This is not a sign of poor data quality; rather, it accurately reflects the archaeological reality that these features were infrequently observed during the surface survey. The presence of `NULL`s in this database, in stark contrast to their complete absence in the other databases, represents a superior and more honest approach to data modeling. It makes the data safer for analysis, as statistical functions in modern software will correctly ignore `NULL` values by default, preventing the kind of data corruption that is a significant risk with the sentinel value approach. This finding confirms that the `text_nulls` model successfully implements a modern, robust standard for handling missing data.

### 7.3. Data Complexity: Cardinality Analysis

#### 7.3.1. Data & Methods

This section analyzes the complexity of the data within each column by measuring its **cardinality**. The data is sourced from the `tmp_benchmark_wide_text_nulls_column_profiles.json` file. Cardinality is defined as the number of unique or distinct values present in a column. This metric is a fundamental characteristic of a dataset and is critical for understanding the nature of each attribute.

The analysis of cardinality helps to distinguish between different types of columns. Columns with very high cardinality are typically identifiers or raw counts. Columns with very low cardinality (e.g., 2 to 10 distinct values) usually represent categorical variables, flags, or status codes. By examining the distribution of cardinalities in this flattened table, we can understand the data complexity and the direct impact of converting numeric codes to descriptive text.

In [ ]:
display_header("Column Cardinality Distribution")

if column_profiles_df is not None and not column_profiles_df.empty:
    # Create a full column identifier
    df = column_profiles_df.copy()
    df["full_column_name"] = (
        df["tablename"].astype(str) + "." + df["column_name"].astype(str)
    )

    # Display tables of highest and lowest cardinality columns
    display(Markdown("**Columns with Highest Cardinality (Most Unique)**"))
    display(
        df.nlargest(10, "distinct_values_estimate")[
            ["full_column_name", "distinct_values_estimate"]
        ]
    )

    display(Markdown("**Columns with Lowest Cardinality (Least Unique)**"))
    display(
        df[df["distinct_values_estimate"] > 1].nsmallest(
            10, "distinct_values_estimate"
        )[["full_column_name", "distinct_values_estimate"]]
    )

    # Create a histogram of cardinalities to see the distribution
    fig = px.histogram(
        df,
        x="distinct_values_estimate",
        log_y=True,
        title=f"Distribution of Column Cardinalities in {DATABASE_NAME}",
        labels={"distinct_values_estimate": "Number of Distinct Values (Cardinality)"},
    )
    fig.show()
else:
    print("No column profile data available.")

### Column Cardinality Distribution

**Columns with Highest Cardinality (Most Unique)**

**Columns with Lowest Cardinality (Least Unique)**

#### 7.3.2. Results & Discussion

The cardinality analysis of the `tmp_benchmark_wide_text_nulls` database reflects a well-structured analytical table. The `SSN` column correctly shows the highest cardinality with 5,050 unique values, serving as the primary identifier. Raw count variables, such as `obsidianTot` (386 unique values), exhibit high cardinality as expected.

The key insight from this analysis comes from the low-cardinality columns. The histogram is heavily skewed towards the low end, with dozens of columns having fewer than 10 unique values. These represent the categorical variables that were successfully converted from integer codes to text. For example, `boundInfoQual` has a cardinality of 3, corresponding to its descriptive text values (e.g., `"1. Good"`), and the pivoted boolean personnel columns (e.g., `fieldWorkers_Millon_R`) have a cardinality of 2 (`True`/`False`). This profile is ideal for analysis: high-cardinality identifiers provide unique keys, high-cardinality numeric fields provide quantitative data, and low-cardinality text fields provide ready-to-use, human-readable categorical variables. The cardinality distribution confirms that the database structure is not only performant but also semantically clear and optimized for direct use in statistical and machine learning software.

---
## 8. Performance Benchmark Analysis

### 8.1. Data, Theory & Methods

This section evaluates the analytical performance of the `tmp_benchmark_wide_text_nulls` database by measuring query execution latency on a set of standardized, canonical queries. The data for this analysis is sourced from the `tmp_benchmark_wide_text_nulls_performance_benchmarks.csv` file, which logs the results of these benchmark tests. The methodology is designed to provide a fair, reproducible, and representative test of the schema's efficiency under different analytical workloads.

The methodology involves executing a predefined set of three canonical queries against the database and measuring the time taken for each to complete, reported in milliseconds. These queries, stored in `phases/01_LegacyDB/sql/canonical_queries/canonical_queries_benchmark.sql`, are designed to test the performance of the single-table, denormalized architecture:
1.  **Baseline Scan:** A simple `COUNT(*)` on the `wide_format_data` table to establish a baseline for raw I/O performance.
2.  **Join Performance (Simulated):** A query that would require a complex join in a normalized schema (retrieving and ordering by obsidian counts) is performed as a simple `SELECT` and `ORDER BY` operation here. This directly tests the benefit of having pre-joined the data.
3.  **Complex Filtering:** A query that filters on multiple attributes (site unit and collection year) and performs an aggregation, a common analytical task that would require a multi-table join in a normalized schema.

By analyzing the latency of these queries, we can quantitatively measure the performance advantages of the denormalized design, directly testing the central hypothesis of this report.

In [ ]:
display_header("Canonical Query Performance Results")

if performance_df is not None and not performance_df.empty:
    display(performance_df[["query_name", "latency_ms", "status"]])

    # Plot the results for successful queries
    success_df = performance_df[performance_df["status"] == "Success"]
    if not success_df.empty:
        fig = px.bar(
            success_df,
            x="query_name",
            y="latency_ms",
            title=f"Query Latency for {DATABASE_NAME}",
            labels={"latency_ms": "Latency (ms)", "query_name": "Canonical Query"},
        )
        fig.show()
else:
    print("No performance benchmark data available.")

### Canonical Query Performance Results

### 8.2. Results: Query Performance for `tmp_benchmark_wide_text_nulls`

The performance benchmark results for the `tmp_benchmark_wide_text_nulls` database confirm the outstanding efficiency of the wide-format, denormalized architecture. All three canonical queries executed successfully with very low latencies:

*   **Baseline Performance - Query 1.1:** 2.03 ms
*   **Join Performance - Query 2.1:** 16.12 ms
*   **Complex Filtering - Query 3.1:** 1.94 ms

The baseline scan established a rapid I/O benchmark of approximately 2 ms. The most complex query, which simulates a join by ordering a large dataset, executed in only **16.12 ms**. This represents a minimal performance degradation of only **7.9 times** compared to the baseline scan. The complex filtering query was exceptionally fast, executing in **1.94 ms**, a time comparable to the baseline itself. These results indicate a highly consistent and performant profile across different analytical workloads, with even the most demanding query executing in a fraction of the time required by the legacy normalized schemas.

### 8.3. Discussion: Impact of Schema on Analytical Performance

The performance benchmark results provide decisive, quantitative validation for the hypothesis that a denormalized, single-table architecture is vastly superior for analytical workloads. The extremely low and consistent latencies are a direct result of completely eliminating the need for relational join operations at query time. By pre-joining all data tables and lookup tables during the ETL process, all data for a given site is stored contiguously, allowing the database engine to satisfy complex queries through highly-efficient table scans and filtering operations.

This finding is the cornerstone of the Phase 1 architectural recommendation. It proves empirically that the significant computational overhead of navigating complex, multi-table schemas like `TMP_DF9` (12.1x slowdown) or even partitioned schemas like `TMP_DF8` (8.3x slowdown) can be eliminated by adopting a wide-format design. The `tmp_benchmark_wide_text_nulls` database successfully demonstrates that for a read-heavy, static analytical dataset like the TMP, sacrificing normalization principles for a denormalized structure yields a dramatic and essential improvement in performance. This benchmark, with its superior usability and data integrity, serves as the definitive proof-of-concept for the proposed Phase 2 architecture.

---
## 9. Final Report and Recommendations

### 9.1. Summary of Results

This analysis provides a comprehensive profile of the `tmp_benchmark_wide_text_nulls` database, confirming its role as the ideal architectural model, demonstrating superior performance, usability, and data integrity.

*   **Structural Complexity:**
    The database exhibits the lowest possible structural complexity. It consists of a **single table** (`wide_format_data`) with **337 columns**, and its relational complexity metrics (JDI, NF) are zero. The ERD is a single, isolated box, visually representing this maximal simplicity. The schema's composition is a mix of `text` (**36.2%**), `double precision` (**47.8%**), and `boolean` (**15.7%**) columns, reflecting a deliberate design choice to store categorical data in a human-readable format.

*   **Data Quality & Health Concerns:**
    This benchmark demonstrates a modern, robust approach to data quality. It is the only database in the analysis that correctly uses standard **SQL `NULL`s** to represent missing data, a critical feature for ensuring data integrity. The presence of `NULL`s in **265 columns**, with high percentages in rare artifact categories, accurately reflects the nature of the archaeological record. The table health is questionable, with an exceptionally high bloat of **142.98%**, though this is an artifact of the ETL process in a static table and not an operational concern.

*   **Performance Profile:**
    The performance benchmarks demonstrated the outstanding efficiency of the denormalized design. All queries executed with very low latency. A simple baseline scan took **2.03 ms**, and the most complex analytical query took only **16.12 ms**. This represents a minimal performance degradation of only **7.9 times** over the baseline, proving the architecture is consistently fast across different analytical workloads.

### 9.2. Discussion

The most striking feature of the `tmp_benchmark_wide_text_nulls` database is its successful synthesis of **high performance with high usability and data integrity**. It is the only database in the Phase 1 analysis that solves all three major architectural challenges simultaneously. Its primary characteristic is that of a modern, analysis-ready data warehouse table.

By denormalizing the structure into a single table, it achieves the exceptional performance of its `numeric` twin, eliminating the join-related bottlenecks of the legacy schemas. However, it surpasses the `numeric` benchmark by taking two crucial extra steps defined in its `flatten_df9_text_nulls.sql` ETL script:
1.  It replaces opaque integer codes with human-readable text, making the data immediately interpretable.
2.  It replaces non-standard sentinel values with industry-standard SQL `NULL`s, protecting against analytical errors.

This database therefore represents the complete architectural vision for the Phase 2 redesign: a schema that is not only fast but also safe, clear, and easy to use.

### 9.3. Conclusions & Implications for Phase 2 Redesign

Based on its superior performance, usability, and data integrity, this report concludes that the `tmp_benchmark_wide_text_nulls` database is the definitive model for the Phase 2 redesign. It successfully addresses the critical flaws of the legacy systems and provides a robust, modern, and efficient foundation for future analysis.

*   **Based on this analysis, what are the key strengths and weaknesses of this database's design?**
    *   **Strengths:**
        *   **Exceptional Performance:** The single-table, denormalized structure is extremely fast, eliminating the severe join penalties of normalized schemas.
        *   **Superior Usability:** The use of descriptive `text` for categorical data makes the database immediately human-readable and easy to analyze without external codebooks.
        *   **High Data Integrity:** The correct use of standard SQL `NULL`s for missing data prevents the corruption of statistical analyses and aligns with modern best practices.
    *   **Weaknesses:**
        *   **Data Redundancy:** The primary theoretical weakness is data redundancy (e.g., repeating a text description for a code). However, this is irrelevant for a static, read-only analytical dataset like the TMP, where data modification anomalies are not a risk.
        *   **Larger File Size:** The use of text fields results in a larger on-disk file size (62 MB) compared to a numerically-coded version. This is a minor and acceptable trade-off for the immense gains in usability and data safety.

*   **What specific aspects of this schema should be preserved, changed, or discarded in the final unified database?**
    *   **Preserve:**
        *   **The Single-Table, Denormalized Architecture:** This is the core principle and must be preserved. It is the key to achieving the project's performance goals.
        *   **Human-Readable Text for Categories:** This is a critical usability feature and must be preserved to ensure the database is accessible to its target audience of archaeologists.
        *   **Standard SQL `NULL`s for Missing Data:** This is a non-negotiable data integrity feature that must be the standard for the final unified database.
    *   **Change:**
        *   No aspects of this benchmark's *design philosophy* need to be changed. The ETL process for creating the final database will incorporate this model while also drawing on data cleaning and content from other legacy sources.
    *   **Discard:**
        *   This benchmark serves as the target model. It represents the *result* of discarding the weaknesses of the legacy systems: the structural fragmentation, the use of sentinel values, and the reliance on opaque numeric codes.